File: infiniteCampusAPI.py
from datetime import date
import requests

TOKEN_URL = "https://iacloud2.infinitecampus.org/campus/oauth2/token?appName=***********"
CLIENT_ID = "***********"
CLIENT_SECRET = "***********"
API_URL = "https://iacloud2.infinitecampus.org/campus/api/oneroster/v1p2/***********/ims/oneroster/"

def main():
    birthdayIDs = getBirthdayIDs(date.today())
    names = getNamesFromID(birthdayIDs)
    print(names)


# Pulls data from local Davis WeatherLink sensor
def getBirthdayIDs(givenDate: date) -> list[str]:
    try:
        url = API_URL + "rostering/v1p2/demographics"
        parameters = {
            "fields": "sourcedId,birthDate",
            "limit": 5000
        }
        headers = {"Authorization": "Bearer " + getOAuthToken}
        response = requests.get(url, parameters, headers=headers)

        dateStr = givenDate.strftime("%m-%d")
        birthdayIDs = []
        for responseDict in response.json()["demographics"]:
            if ("birthDate" in responseDict and "sourcedId" in responseDict
                and responseDict["birthDate"][5:] == dateStr):
                birthdayIDs.append(responseDict["sourcedId"])
        return birthdayIDs
    except Exception as e:
        print("Error with request: ", e)
        return ["ERROR REQUESTING DEMOGRAPHICS DATA"]

def getNamesFromID(ids: list[str]) -> list[str]:
    try:
        url = API_URL + "rostering/v1p2/users/"
        parameters = {
            "fields": "givenName,familyName", # TODO: limit to students
            "limit": 1
        }
        headers = {"Authorization": "Bearer " + getOAuthToken} # TODO: Use existing token

        names = []
        for id in ids:
            idURL = url + id
            response = requests.get(idURL, parameters, headers=headers)
            print(response.json())
        return names
    except Exception as e:
        print("Error with request: ", e)
        return ["ERROR REQUESTING USER DATA"]


def getOAuthToken() -> str:
    tokenReqPayload = {"grant_type": "client_credentials"}
    tokenResponse = requests.post(TOKEN_URL, tokenReqPayload, auth=(CLIENT_ID, CLIENT_SECRET))
    if tokenResponse.status_code != 200:
        print("ERROR FETCHING OAUTH2 TOKEN")
        raise Exception("ERROR FETCHING OAUTH2 TOKEN")
    return tokenResponse.json()["access_token"]

if __name__ == "__main__":
    main()

File: infiniteCampusAPI.js
const IC_TOKEN_URL = "https://tokenurl.com";
const IC_CLIENT_ID = "123456789";
const IC_CLIENT_SECRET = "sUpErSeCrEt";
const IC_API_URL = "https://iacloud2.infinitecampus.org/more";
const SCHOOL_ID = "school-string";

const TEMPTOKEN = "usedForTesting";

function getICBirthdays(date) {
    try {
        const token = TEMPTOKEN;
        const birthdayIDs = getBirthdayIDs(date, token);
        const studentNames = getSchoolNamesFromID(birthdayIDs, "students", token);
        const teacherNames = getSchoolNamesFromID(birthdayIDs, "teachers", token);

        return [studentNames, teacherNames];
    }
    catch (err) {
        Logger.log("Error Caught: %s", err.message);
        return ["ERROR REQUESTING DEMOGRAPHICS DATA"];
    }
}

function getBirthdayIDs(date, token) {
    const options = {
        headers: {
            Authorization: "Bearer " + token
        }
    };
    const parameters = {
        fields: "sourcedId,birthDate",
        limit: 5000
    };
    const url = buildUrl_(API_URL + "rostering/v1p2/demographics", parameters);
    const response = UrlFetchApp.fetch(url, options);
    const responseData = JSON.parse(response.getContentText());
    
    const dateStr = Utilities.formatDate(date, "GMT-6", "MM-dd");
    const birthdayIDs = [];
    for (person of responseData["demographics"]) {
        if (person["birthDate"]?.includes(dateStr)) {
            birthdayIDs.push(person["sourcedId"]);
        }
    }
    return birthdayIDs;
}

function getSchoolNamesFromID(ids, group, token) {
    const options = {
        headers: {
            Authorization: "Bearer " + token
        }
    };
    const parameters = {
        filter: "sourcedId='" + ids.join("' OR sourcedId='") + "'",
        fields: "givenName,familyName",
        limit: 1
    };
    const baseUrl = API_URL + "rostering/v1p2/schools/" + SCHOOL_ID + '/' + group;
    const url = buildUrl_(baseUrl, parameters);
    const response = UrlFetchApp.fetch(url, options);
    const responseData = JSON.parse(response.getContentText());
    
    const names = [];
    for (user of responseData["users"]) {
        names.push(user["givenName"] + " " + user["familyName"]);
    }
    return names;
}

function getOAuthToken() {
    const options = {
        payload: {
            grant_type: "client_credentials"
        },
        headers: {
            Authorization: "Basic " + Utilities.base64Encode(IC_CLIENT_ID + ':' + IC_CLIENT_SECRET)
        }
    };
    const response = UrlFetchApp.fetch(IC_TOKEN_URL, options);
    const responseData = JSON.parse(response.getContentText());

    return responseData["access_token"];
}


// GoogleWorkspace OAuth2 Library url builder
// https://github.com/googleworkspace/apps-script-oauth2
function buildUrl_(url, params) {
    var paramString = Object.keys(params).map(function(key) {
        return encodeURIComponent(key) + '=' + encodeURIComponent(params[key]);
    }).join('&');
    return url + (url.indexOf('?') >= 0 ? '&' : '?') + paramString;
}